File: circleDetector.m
%% RESET 
clc;
clear all;

%% LOAD DATA
imageFolder = 'C:\Users\matsk\OneDrive\Pulpit\Notes\VI_SEMESTER\AO\Labs\Lab2\data';

imageFiles = dir(fullfile(imageFolder, '*.bmp')); 
numImages = length(imageFiles);
images = cell(1, numImages);

for k = 1:numImages
    filename = fullfile(imageFolder, imageFiles(k).name);
    images{k} = imread(filename);
end

%% CONVERT TO HSV

fhsvImg = cell(numImages, 1);
hue = cell(numImages, 1);
sat = cell(numImages, 1);
val = cell(numImages, 1);

for k = 1:numImages
    fhsvImg{k} = rgb2hsv(images{k});
    hue{k} = fhsvImg{k}(:,:,1);
    sat{k} = fhsvImg{k}(:,:,2);
    val{k} = fhsvImg{k}(:,:,3);
end

%% DETECT RED REGION & MEASURE AREA

redMask = cell(numImages, 1);       % combined masks
cleanMask = cell(numImages, 1);     % cleaned binary masks
areaPixels = zeros(numImages, 1);   % area values

for k = 1:numImages
    % threshold
    redMask1 = (hue{k} >= 0.0 & hue{k} <= 0.05) & sat{k} > 0.35 & val{k} > 0.15;
    redMask2 = (hue{k} >= 0.95 & hue{k} <= 1.0) & sat{k} > 0.35 & val{k} > 0.15;
    redMask{k} = redMask1 | redMask2;

    % fill the holes
    maskFilled = imfill(redMask{k}, 'holes');
    maskCleaned = bwareaopen(maskFilled, 100); 
    cleanMask{k} = maskCleaned;

    % measure area in pixels
    areaPixels(k) = sum(maskCleaned(:));
end

%% SHOWING RESULTS

for k = 1:numImages
    figure('Name', ['Image ', num2str(k)], 'NumberTitle', 'off');

    % original img
    subplot(1,2,1);
    imshow(images{k});
    title('Original');

    % overlayed img
    subplot(1,2,2);
    imshow(images{k});
    hold on;

    % blue overlay
    blueOverlay = cat(3, zeros(size(cleanMask{k})), zeros(size(cleanMask{k})), ones(size(cleanMask{k})));
    h = imshow(blueOverlay);
    set(h, 'AlphaData', 0.4 * cleanMask{k});

    title(['Overlayed (Blue Mask) — Area: ', num2str(areaPixels(k)), ' px']);
    hold off;
end

%% SAVE RESULTS

imageNames = {imageFiles.name}';
resultTable = table(imageNames, areaPixels, 'VariableNames', {'Filename', 'RedAreaPixels'});
disp(resultTable);


%% BAR CHART

figure;
bar(resultTable.RedAreaPixels, 'FaceColor', [0.2 0.6 1]);
xticks(1:height(resultTable));
xticklabels(resultTable.Filename);
xtickangle(45);
ylabel('Red Area [pixels]');
xlabel('Image Filename');
title('Red-Filled Circle Area per Image');
grid on;

%% LINE PLOT

figure;
plot(resultTable.RedAreaPixels, '-o', 'LineWidth', 2, 'Color', [0.1 0.4 0.8]);
xticks(1:height(resultTable));
xticklabels(resultTable.Filename);
xtickangle(45);
ylabel('Red Area [pixels]');
xlabel('Image Filename');
title('Trend of Red Area across Images');
grid on;

%% PIE CHART

figure;
pie(resultTable.RedAreaPixels, resultTable.Filename);
title('Proportional Red Area per Image');

File: circleDetector.py
import os
import cv2
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# === LOAD IMAGES ===

image_folder = r'C:\Users\matsk\PyCharmMiscProject\AOLab2\data'
image_files = [f for f in os.listdir(image_folder) if f.lower().endswith('.bmp')]
images = [cv2.imread(os.path.join(image_folder, f)) for f in image_files]
num_images = len(images)

# === CONVERT TO HSV ===

hsv_images = [cv2.cvtColor(img, cv2.COLOR_BGR2HSV) for img in images]
hue = [img[:, :, 0] / 179.0 for img in hsv_images]
sat = [img[:, :, 1] / 255.0 for img in hsv_images]
val = [img[:, :, 2] / 255.0 for img in hsv_images]

# === DETECT RED REGIONS & MEASURE AREA ===

clean_masks = []
areas = []

for i in range(num_images):
    h, s, v = hue[i], sat[i], val[i]

    # thresholds
    red_mask1 = ((h >= 0.0) & (h <= 0.05)) & (s > 0.35) & (v > 0.15)
    red_mask2 = ((h >= 0.95) & (h <= 1.0)) & (s > 0.35) & (v > 0.15)
    combined_mask = np.logical_or(red_mask1, red_mask2).astype(np.uint8)

    # fill holes
    mask_filled = cv2.morphologyEx(combined_mask, cv2.MORPH_CLOSE, np.ones((5, 5), np.uint8))
    mask_cleaned = cv2.morphologyEx(mask_filled, cv2.MORPH_OPEN, np.ones((5, 5), np.uint8))
    mask_cleaned = cv2.connectedComponentsWithStats(mask_cleaned, 8)[1] > 0  # Clean tiny blobs

    clean_masks.append(mask_cleaned)
    areas.append(np.sum(mask_cleaned))

# === SHOW RESULTS WITH OVERLAY ===

for i in range(num_images):
    img_rgb = cv2.cvtColor(images[i], cv2.COLOR_BGR2RGB)

    plt.figure(figsize=(10, 5))

    # original
    plt.subplot(1, 2, 1)
    plt.imshow(img_rgb)
    plt.title('Original')
    plt.axis('off')

    # overlay
    plt.subplot(1, 2, 2)
    overlay = img_rgb.copy()
    overlay[clean_masks[i]] = [0, 0, 255]
    blended = cv2.addWeighted(img_rgb, 0.6, overlay, 0.4, 0)
    plt.imshow(blended)
    plt.title(f'Overlayed — Area: {areas[i]} px')
    plt.axis('off')

    plt.tight_layout()
    plt.show()

# === TABLE OF RESULTS ===

df = pd.DataFrame({
    'Filename': image_files,
    'RedAreaPixels': areas
})
print(df)

# === BAR CHART ===

plt.figure(figsize=(10, 6))
plt.bar(df['Filename'], df['RedAreaPixels'], color='skyblue')
plt.xticks(rotation=45, ha='right')
plt.ylabel('Red Area [pixels]')
plt.xlabel('Image Filename')
plt.title('Red-Filled Circle Area per Image')
plt.grid(True)
plt.tight_layout()
plt.show()

# === LINE PLOT ===

plt.figure(figsize=(10, 6))
plt.plot(df['RedAreaPixels'], '-o', color='dodgerblue', linewidth=2)
plt.xticks(ticks=range(num_images), labels=df['Filename'], rotation=45, ha='right')
plt.ylabel('Red Area [pixels]')
plt.xlabel('Image Filename')
plt.title('Trend of Red Area across Images')
plt.grid(True)
plt.tight_layout()
plt.show()

# === PIE CHART ===

plt.figure(figsize=(8, 8))
plt.pie(df['RedAreaPixels'], labels=df['Filename'], autopct='%1.1f%%', startangle=140)
plt.title('Proportional Red Area per Image')
plt.tight_layout()
plt.show()
